pallets/fragments/src/lib.rs

```diff
@@ -492,6 +593,169 @@ pub mod pallet {
 			Ok(())
 		}
 
+		/// **Alters** the **custom metadata** of a **Fragment Definition** (whose ID is `fragment_hash`) by **adding or modifying a key-value pair** (`metadata_key.clone`,`blake2_256(&data.encode())`)
+		/// to the **BTreeMap field `custom_metadata`** of the **existing Fragment Definition's Struct Instance**.
+		/// Furthermore, this function also indexes `data` in the Blockchain's Database and stores it in the IPFS
+		///
+		/// # Arguments
+		///
+		/// * `origin` - The origin of the extrinsic / dispatchable function
+		/// * `fragment_hash` - **ID of the Fragment Definition**
+		/// * `metadata_key` - The key (of the key-value pair) that is added in the BTreeMap field `custom_metadata` of the existing Fragment Definition's Struct Instance
+		/// * `data` - The hash of `data` is used as the value (of the key-value pair) that is added in the BTreeMap field `custom_metadata` of the existing Fragment Definition's Struct Instance
+		#[pallet::weight(50_000)]
+		pub fn set_definition_metadata(
+			origin: OriginFor<T>,
+			// fragment hash we want to update
+			fragment_hash: Hash128,
+			// Think of "Vec<u8>" as String (something to do with WASM - that's why we use Vec<u8>)
+			metadata_key: Vec<u8>,
+			// data we want to update last because of the way we store blocks (storage chain)
+			data: Vec<u8>,
+		) -> DispatchResult {
+
+			let who = ensure_signed(origin)?;
+
+			let proto_hash =
+				<Definitions<T>>::get(fragment_hash).ok_or(Error::<T>::NotFound)?.proto_hash; // Get `proto_hash` from `fragment_hash`
+			let proto: Proto<T::AccountId, T::BlockNumber> =
+				<Protos<T>>::get(proto_hash).ok_or(Error::<T>::ProtoNotFound)?;
+			let proto_owner: T::AccountId = match proto.owner { // Get `proto_owner` from `proto`
+				ProtoOwner::User(owner) => Ok(owner),
+				_ => Err(Error::<T>::ProtoOwnerNotFound),
+			}?;
+			ensure!(who == proto_owner, Error::<T>::NoPermission); // Ensure `who` is `proto_owner`
+
+			// TO REVIEW
+			ensure!(!<DetachedHashes<T>>::contains_key(&proto_hash), Error::<T>::Detached); // Ensure `proto_hash` isn't detached
+
+			let data_hash = blake2_256(&data);
+
+			// we need this to index transactions
+			let extrinsic_index = <frame_system::Pallet<T>>::extrinsic_index()
+				.ok_or(Error::<T>::SystematicFailure)?;
+
+			// Write STATE from now, ensure no errors from now...
```

Comment:
Well it better be clear...
Substrate pallets are not safe like smart contracts, everything you write INTO state will stick even if your code MISERABLY FAILS 😄 so you need to ensure no errors, it can re roll transactions but it's not explicit/easy/automatic.